Improve Windows builds

Archomeda opened the pull request:
As I've mentioned in #1, I've been playing around with windeployqt. Not only have I applied that command, but also some other things:

- Automatic generation of the compiled translation files (it currently happens when running `qmake`). Sadly I can't seem to make it better than to include it as a `system` command within `src.pro`. I tried using `QMAKE_EXTRA_COMPILERS`, but for some reason qmake adds duplicate filters that Visual Studio doesn't like. Also, I still believe that the language files are compiled within the executable as a resource. This is the reason why I removed the translations folder from the output completely.
- I also removed the precompiled `*.qm` files as they are no longer needed.
- With the usage of windeployqt, vcredist_x64.exe or vcredist_x86.exe should be copied automatically as well. This removes the need of copying msvcr120.dll and msvcp120.dll manually. I've added an additional installation step in the installer to support it.
- I changed the `QMAKE_POST_LINK` to also support windeployqt. Apparently, windeployqt doesn't add the ICU DLLs if I don't specify it explicitly. It still runs fine on my pc, and also on my laptop that doesn't have Qt installed.

Can you test it? I hope it builds and works correctly on your end. If it doesn't, I'll see what I can do.

psieg commented:
ok to test

Archomeda commented:
Hmm... it seems the redistributable package cannot be found in the proper directory. This is odd.

```
C:\jenkins\workspace\Prismatik_Windows\Software\src\bin\Prismatik.exe 64 bit, release executable
Skipping plugin qsvgicon.dll due to disabled dependencies.
Skipping plugin qsvg.dll due to disabled dependencies.
Direct dependencies: Qt5Core Qt5Gui Qt5Network Qt5SerialPort Qt5Widgets
All dependencies   : Qt5Core Qt5Gui Qt5Network Qt5SerialPort Qt5Widgets
To be deployed     : Qt5Core Qt5Gui Qt5Network Qt5SerialPort Qt5Widgets
EXEC : warning : Cannot find Visual Studio redistributable in C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\redist. [C:\jenkins\workspace\Prismatik_Windows\Software\src\Prismatik.vcxproj]
```

psieg commented on Software/dist_windows/script.iss:
```diff
@@ -67,11 +64,9 @@ Source: "content/prismatik-unhook.dll"; DestDir: "{app}"; MinVersion: 6.1.7600;
-Source: "content/msvcr120.dll"; DestDir: "{app}"; Flags: ignoreversion
-Source: "content/msvcp120.dll"; DestDir: "{app}"; Flags: ignoreversion
+; Visual Studio redistributable package
+Source: "content/vcredist_x64.exe"; DestDir: {tmp}; Flags: deleteafterinstall
```
The problem with using the redist installers is the lack of a dependency between the entries in the control panel. Since we aren't using MSI installers we can't use the preferred version, merge modules, to ship the runtime. This is why i chose to include the dlls directly. It's also why the 32bit files are statically linked now - to avoid having to ship both runtimes

see http://blogs.msdn.com/b/vcblog/archive/2007/10/12/how-to-redistribute-the-visual-c-libraries-with-your-application.aspx

Archomeda replied:
I see, I did not know that. Windows has a tendency for having multiple ways to achieve something, and everyone recommends something else. The Qt team thought this was the best idea, but it's always possible to deviate from that. I'll edit it back. Do you prefer the previous one where you copied it from the Windows directory, or the one that you can find in `<Visual Studio install dir>\VC\redist\<arch>\Microsoft.VC120.CRT`?

psieg replied:
Probably the latter would be better IMO

psieg commented on Software/src/src.pro:
```diff
@@ -104,34 +107,13 @@ win32 {
+        $$[QT_INSTALL_BINS]/windeployqt --no-angle --no-svg --no-translations \"$(TargetName)$(TargetExt)\"$$escape_expand(\r\n)\
+        del opengl32sw.dll$$escape_expand(\r\n)\
+        rd /s /q bearer $$escape_expand(\r\n)\
+        rd /s /q imageformats $$escape_expand(\r\n)\
```
do you know what ``bearer`` and ``imageformats`` do? The current release does not include them and seems to work fine...

Archomeda replied:
These are extra Qt plugins apparently. `bearer` has something to do with [connectivity management](http://doc.qt.io/qt-5/bearer-management.html). I think this might have been copied because of the dependency on `QtNetwork`. `imageformats` adds a lot of DLLs mentioning various image formats, but I'm pretty sure Prismatik doesn't use them. Otherwise my install should not be working as Prismatik uses icons etc.

psieg replied:
I read this wrong, you are using windows commands to remove them :+1: I thought you were re-adding them.

Archomeda replied:
Ahh, makes sense. But yeah, I decided to rewrite those commands to native Windows ones, this way people don't need to have MinGW or any other posix environment installed in their PATH.

psieg commented on Software/dist_windows/script32.iss:
```diff
@@ -57,18 +57,13 @@ Source: "content/Qt5Network.dll"; DestDir: "{app}"; Flags: ignoreversion
 Source: "gpl-3.0.txt"; DestDir: "{app}"; Flags: ignoreversion
-Source: "content/icudt54.dll"; DestDir: "{app}"; Flags: ignoreversion
-Source: "content/icuin54.dll"; DestDir: "{app}"; Flags: ignoreversion
-Source: "content/icuuc54.dll"; DestDir: "{app}"; Flags: ignoreversion
```
windeployqt says the icu* dlls are not needed? Last time I tried Qt5Core had a hard dependency on them

Archomeda replied:
I can force it to copy it I think, by adding a command line argument. But even according to Dependency Walker, Qt5Core doesn't seem to have a hard dependency on it.

![naamloos](https://cloud.githubusercontent.com/assets/1467127/8908908/eb8f2138-347e-11e5-8bd1-624880c8f03a.png)

psieg replied:
That's wierd, have you tested an installation on a PC without Qt? According to  http://doc.qt.io/qt-5/windows-deployment.html

>Qt relies on the ICU library for unicode support. You must include the ICU DLLs that are located in the bin directory of your Qt installation if Qt was configured to use ICU. The Qt version bundled in the Qt5 package uses ICU, so deployment is needed there. The ICU DLLs are version dependent and have to match the ones your Qt version was linked against.

Archomeda replied:
I've tested it on a clean Windows 10 VM that I installed a few days ago. Got similar results:

![windows 10-2015-07-27-17-50-42](https://cloud.githubusercontent.com/assets/1467127/8910740/bbc378a0-3488-11e5-8d14-a6000f467426.png)

There is a couple of missing stuff though, but this is related to the system files itself. I'm not sure why, and Prismatik seems to start up fine.

Although I am encountering something odd. I have no clue why, but it should be displaying the Smoothness setting instead of those weird artifacts. Maybe it's because I've used a virtual device instead of a Lightpack?
![windows 10-2015-07-27-17-58-33](https://cloud.githubusercontent.com/assets/1467127/8910816/3cadeacc-3489-11e5-9eb3-07f0d642e4f4.png)

Archomeda replied:
To add on that, it would seem that Qt decided to change their mind since 5.4 regarding ICU: http://wiki.qt.io/Qt-5-QLocale#Platform_Support.

>WIndows 7 and later provides advanced localization functions clearly based on CLDR and broadly comparable to ICU. Windows Vista lacks many of these new API relying mainly on Win32 API's, but with a < 2% market share can be safely ignored provided sensible fallbacks occur. Windows XP only has Win32 API which lack most of the required advanced functions, but basic functionality can still be supported using some fallbacks. The issue comes with what platform Qt is compiled for, runtime checks may not be sufficient to work around missing API calls.

So I guess they use the Windows API natively now, instead of using the ICU libraries?

psieg replied:
I found [this](http://stackoverflow.com/a/22449478) and came to a similar conclusion, so lets try and leave them out.

The "artifacts" are win10 themed sliders, so thats all right.

Archomeda replied:
If it works out fine, it will save 50% of the installer file size. Which is also great.

No, I meant the black rectangles. But I seem to recall that these also showed up when I added a virtual device months ago. So I guess it's just small bug in the UI.

psieg replied:
Very nice.
The rectangles are your virtual LED device. If you move the window to a screen border some of them should turn white

Archomeda replied:
Oh derp, now that you mention it :sweat:

Archomeda commented:
I'll add a newer commit soon that improves the build-vars.prf, etc.

Archomeda commented:
My last commit changes the way how you use qmake. You no longer have to create `build-vars.prf` (although you still can). The default values are automatically loaded. You can overwrite the defaults by either using `build-vars.prf` or setting it as an environment variable. I also removed obsolete variables.

- `QTDIR`: Removed as it was only used in some areas that are now rewritten as of this PR
- `DIRECTX_SDK_DIR`: Can be set from the environment or from `build-vars.prf`
- `WIN81_SDK_DIR`: Was not even used by the script, it was only used to check if this SDK was to be used instead of DirectX
- `QMAKE_TARGET.arch`: Changed to `TARGET_ARCH` variable, default is now `x86_64`, can be overwritten from the environment or from `build-vars.prf`

psieg commented:
add to whitelist

psieg commented on Software/src/src.pro:
```diff
@@ -104,34 +107,14 @@ win32 {
+        if exist imageformats ( rd /s /q imageformats ) $$escape_expand(\r\n)\
+        copy /y \"$(VcInstallDir)redist\\$(PlatformTarget)\\Microsoft.VC$(PlatformToolsetVersion).CRT\\msvcr$(PlatformToolsetVersion).dll\" .\ $$escape_expand(\r\n)\
+        copy /y \"$(VcInstallDir)redist\\$(PlatformTarget)\\Microsoft.VC$(PlatformToolsetVersion).CRT\\msvcp$(PlatformToolsetVersion).dll\" .\ $$escape_expand(\r\n)\
+        copy /y ..\\..\\lib\\libraryinjector.dll .\\
```
Since all the other dlls copy themselves to src/bin, the libraryinjector should too. Or, even better, forget about the lib folder and compile everything to bin (not src/bin)

Archomeda replied:
It should be copied automatically yeah, but it's not happening for some reason. I have to check what's happening.

psieg replied:
[offsetfinder.pro](https://github.com/psieg/Lightpack/blob/master/Software/offsetfinder/offsetfinder.pro) doesn't have the copy-lines the [other project files](https://github.com/psieg/Lightpack/blob/master/Software/hooks/hooks.pro#L47) have, should be a simple thing

Archomeda replied:
I noticed it. I changed the output directories accordingly in the last commits. I also reverted the somewhat duplicated `QMAKE_POST_LINK` as I just noticed they were separated for Visual Studio and MinGW compilers. I'm unsure if MinGW actually works though, since there will be other runtime libraries.

psieg commented:
I'd do it myself but I can't push to your PR: the .pro's have a `DESTDIR` which you can set, then the copy step can be entirely eliminated

Archomeda commented:
>Correction: compile libs to lib but dlls to bin

Whoops, I read this wrong and I thought that you meant all the dlls to lib and only the exe to bin.

psieg commented on Software/src/qtsingleapplication/common.pri:
```diff
@@ -1,4 +1,4 @@
-exists(config.pri):infile(config.pri, SOLUTIONS_LIBRARY, yes): CONFIG += qtsingleapplication-uselib
+#exists(config.pri):infile(config.pri, SOLUTIONS_LIBRARY, yes): CONFIG += qtsingleapplication-uselib
```
I'm not sure what this did but if it's really unnecessary please remove the line

Archomeda replied:
When running `qmake`, this throws an error that it cannot find `config.pri`. Which is weird, since it's checked before trying to include it by using `exists`.

The reason why I commented it out, is because I moved my personal builds to AppVeyor, and [AppVeyor apparently doesn't like output in stderr](https://ci.appveyor.com/project/Archomeda/lightpack/build/5.11.2.1-1/job/7gdthqrlvxwalg3f). So this line prevented the build from continuing. And because this line was originally from the qtsingleapplication source, I decided to not remove it. I will take a look if I can somehow make it working the way it was supposed to.

Archomeda replied:
I can't solve it. For some reason qmake ignores `exist` for `infile`. I can either leave it this way, or add an empty `config.pri`. What do you prefer?

psieg replied:
I can reproduce the problem (though the missing file is not a critical error). From what I can see in the original repo a config.pri was never used and compilation fails if that line is evaluated. I'd say the options are
* remove the line entirely, since it's of no use
* leave it there in case someone needs it sometime. In that case add a comment why it's disabled
I don't mind

psieg commented:
Looks great. Will merge once the singleapplication issue is resolved

Archomeda commented:
Done. I've added a comment that clarifies why it's commented out.
